File: Raven.Client.Lightweight/Linq/AggregationQuery.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using Raven.Abstractions.Data;

namespace Raven.Client.Linq
{
	public class AggregationQuery<T> : AggregationQuery
	{
		public List<Expression<Func<T, bool>>> Ranges { get; set; }

		public static IEnumerable<Facet> GetFacets(List<AggregationQuery<T>> aggregationQueries)
		{
			var facetsList = new List<Facet>();

			foreach (var aggregationQuery in aggregationQueries)
			{
				if (aggregationQuery.Aggregation == FacetAggregation.None)
					throw new InvalidOperationException("All aggregations must have a type");

				var shouldUseRanges = aggregationQuery.Ranges != null && aggregationQuery.Ranges.Count > 0;

				List<string> ranges = null;
				if (shouldUseRanges)
					ranges = aggregationQuery.Ranges.Select(Facet<T>.Parse).ToList();

				var mode = shouldUseRanges ? FacetMode.Ranges : FacetMode.Default;
				facetsList.Add(new Facet
				{
					Name = aggregationQuery.Name,
					DisplayName = aggregationQuery.DisplayName,
					Aggregation = aggregationQuery.Aggregation,
					AggregationType = aggregationQuery.AggregationType,
					AggregationField = aggregationQuery.AggregationField,
					Ranges = ranges,
					Mode = mode
				});
			}
			return facetsList;
		}
	}

	public class AggregationQuery
	{
		public string Name { get; set; }
		public string DisplayName { get; set; }
		public string AggregationField { get; set; }
		public string AggregationType { get; set; }
		public FacetAggregation Aggregation { get; set; }

		public static List<Facet> GetFacets(List<AggregationQuery> aggregationQueries)
		{
			var facetsList = new List<Facet>();

			foreach (var aggregationQuery in aggregationQueries)
			{
				if (aggregationQuery.Aggregation == FacetAggregation.None)
					throw new InvalidOperationException("All aggregations must have a type");

				facetsList.Add(new Facet
				{
					Name = aggregationQuery.Name,
					DisplayName = aggregationQuery.DisplayName,
					Aggregation = aggregationQuery.Aggregation,
					AggregationType = aggregationQuery.AggregationType,
					AggregationField = aggregationQuery.AggregationField,
					Mode = FacetMode.Default
				});
			}
			return facetsList;
		}
	}
}

File: Raven.Client.Java/lib/src/main/java/raven/client/linq/AggregationQuery.java
package raven.client.linq;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;

import raven.abstractions.data.Facet;
import raven.abstractions.data.FacetAggregation;
import raven.abstractions.data.FacetMode;

public class AggregationQuery {

  private String name;
  private String displayName;
  private String aggregrationField;
  private String aggregationType;
  private EnumSet<FacetAggregation> aggregation = EnumSet.noneOf(FacetAggregation.class);

  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  public String getDisplayName() {
    return displayName;
  }

  public void setDisplayName(String displayName) {
    this.displayName = displayName;
  }

  public String getAggregrationField() {
    return aggregrationField;
  }

  public void setAggregrationField(String aggregrationField) {
    this.aggregrationField = aggregrationField;
  }

  public String getAggregationType() {
    return aggregationType;
  }

  public void setAggregationType(String aggregationType) {
    this.aggregationType = aggregationType;
  }

  public EnumSet<FacetAggregation> getAggregation() {
    return aggregation;
  }

  public void setAggregation(EnumSet<FacetAggregation> aggregation) {
    this.aggregation = aggregation;
  }

  public static List<Facet> getFacets(List<AggregationQuery> aggregationQueries) {
    List<Facet> facetsList = new ArrayList<>();
    for (AggregationQuery aggregationQuery : aggregationQueries) {
      if (aggregationQuery.getAggregation().equals(EnumSet.of(FacetAggregation.NONE))) {
        throw new IllegalStateException("All aggregations must have a type");
      }

      Facet facet = new Facet();
      facet.setName(aggregationQuery.getName());
      facet.setDisplayName(aggregationQuery.getDisplayName());
      facet.setAggregation(aggregationQuery.getAggregation());
      facet.setAggregationType(aggregationQuery.getAggregationType());
      facet.setAggregationField(aggregationQuery.getAggregrationField());
      facet.setMode(FacetMode.DEFAULT);

      facetsList.add(facet);
    }
    return facetsList;
  }

}
